db/migrate/20200512044127_change_messages_to_polymorphic.rb

```diff
@@ -0,0 +1,46 @@
+class ChangeMessagesToPolymorphic < ActiveRecord::Migration
+  def up
+    add_column :messages, :messageable_type, :string
+    add_column :messages, :messageable_id, :int
+    Message.all.map do |message|
+      if message.item_id
+        item = message.item
+        message.messageable_type = item.class.name
+        message.messageable_id = item.id
+      elsif message.offer_id
+        offer = message.offer
+        message.messageable_type = offer.class.name
+        message.messageable_id = offer.id
+      elsif message.order_id
+        order = message.order
+        message.messageable_type = order.class.name
+        message.messageable_id = order.id
+      end
+      message.save!(validate: false)
+    end
```

Comment:
Wondering if we want this as a rake task ? I personally like the idea of modifying the data as part of a migration but I know @steveyken has some reservations regarding that 🤔

Reply:
Yep, you're right! :) The main reason is because I've been bitten in the past by cases where the model has changed enough over the years as to render the migration unable to run which immediately halts development and CI specs (if they use `rake db:migrate` instead of `rake db:schema:load`).

Here's a scenario:

One day:
- Create a `Message` model with `super_funky_method`
- Write a migration that utilises `super_funky_method`

Some time later:
- Delete `super_funky_method` from Message model
- Try to run `rake db:migrate`
- Previous migration fails because `Message#super_funky_method` no longer exists

Useful lessons I've learnt:
- Rake migrations that manipulate data should ideally focus on db columns (easiest way to ensure this is by sticking to SQL)
- Don't use any of your own model methods (standard Rails AR methods probably ok)

Should it be a rake task or in a migration? I'd ask myself the following questions:
- Would it be useful to run the code again another time?
- Does this code need to be tested? (much easier to test if abstracted to a standard ruby class). Whilst a TDD purest would always answer yes to this question, I'm open to standard non-complex SQL one-liners inside the migration)
- Does the migration need to trigger other actions that happen in ActiveRecord (lifecycle hooks like `after_save`). If not, SQL code in a migration might be fine.

Bit of a long answer, but thought it worth explaining my thinking behind my reservations.

Reply:
Given the above, here's what I'd do...

```
def up
  add_column :messages, :messageable_type, :string
  add_column :messages, :messageable_id, :int
  execute("UPDATE messages SET messageable_type=''Offer", messageable_id=offer_id WHERE offer_id IS NOT NULL")
  execute("UPDATE messages SET messageable_type=''Item", messageable_id=item_id WHERE item_id IS NOT NULL")
  execute("UPDATE messages SET messageable_type=''Order", messageable_id=order_id WHERE order_id IS NOT NULL")
end
```